illumos-utils/src/opte/illumos/port_manager.rs

```diff
@@ -454,31 +351,239 @@ impl PortManager {
+                Some(SNat4Cfg { external_ip, ports })
+            }
+            None => None,
+        };
+
+        // Describe the external IP addresses for this instance.
+        //
+        // Note that we're currently only taking the first address, which is all
+        // that OPTE supports. The array is guaranteed to be limited by Nexus.
+        // See https://github.com/oxidecomputer/omicron/issues/1467
+        // See https://github.com/oxidecomputer/opte/issues/196
+        let external_ip = match external_ips {
+            Some(ref ips) if !ips.is_empty() => {
+                match ips[0] {
+                    IpAddr::V4(ipv4) => Some(ipv4.into()),
+                    IpAddr::V6(_) => {
+                        return Err(opte_ioctl::Error::InvalidArgument(
+                            String::from("IPv6 is not yet supported for external addresses")
+                        ).into());
+                    }
+                }
+            }
+            _ => None,
+        };
+
+        // TODO-completeness: Add support for IPv6.
+        let vpc_cfg = VpcCfg {
+            ip_cfg: IpCfg::Ipv4(Ipv4Cfg {
+                vpc_subnet: opte_subnet,
+                private_ip,
+                gateway_ip,
+                snat,
+                external_ips: external_ip,
+            }),
+            guest_mac: MacAddr::from(mac.into_array()),
+            gateway_mac: MacAddr::from(gateway.mac.into_array()),
+            vni,
+            phys_ip: self.inner.underlay_ip.into(),
+            boundary_services,
+            // TODO-remove: Part of the external IP hack.
+            //
+            // NOTE: This value of this flag is irrelevant, since the driver
+            // always overwrites it. The field itself is used in the `oxide-vpc`
+            // code though, to determine how to set up the ARP layer, which is
+            // why it's still here.
+            proxy_arp_enable: true,
+            phys_gw_mac: Some(MacAddr::from(
+                self.inner.gateway_mac.into_array(),
+            )),
+            // TODO-completeness (#2153): Plumb domain search list
+            domain_list: vec![],
+        };
+
+        self.create_port_inner(
+            instance_id,
+            PortType::Guest { slot: nic.slot },
+            None,
+            vpc_cfg,
+            firewall_rules,
+            external_ip.is_some(),
+        )
+    }
+
+    pub fn create_svc_port(
+        &self,
+        svc_id: Uuid,
+        svc_name: &str,
+        mac: MacAddr6,
+        external_ip: IpAddr,
+        vni: u32,
+        firewall_rules: &[VpcFirewallRule],
+    ) -> Result<(Port, PortTicket), Error> {
+        // TODO: Assume each service zone has a single port.
+        let port_name = format!("{}_{}0", XDE_LINK_PREFIX, svc_name);
+
+        // TODO: Arbitrary subnet, should come from a "service" VPC.
+        let vpc_subnet: Ipv4Network = "10.0.0.0/24".parse().unwrap();
```

Comment:
this could be wrapped up in a `opte::params::NetworkInterface` too

Reply:
What do you think about this PR defining that service VPC, and creating fake network interfaces in there for services? It would reuse a lot of the existing machinery

Reply:
I deleted another comment and will shove it here instead: if that work is done, users could query for network interfaces in that service VPC to discover the Nexus IP they need to set.

Reply:
Moved a lot of these hardcoded values to a `NetworkInterface` as suggested. Also added subsequent commits to create some built-in orgs/proj/vpc for services to stash these values and use a lot of the existing machinery.

One caveat: these service "NIC"s aren't in the normal `network_interface` table as it currently expects a non-NULL FK to the `instance` table. Tried a different approaches to consolidate but those ended up requiring a lot more significant changes which I'd rather do in a follow-up than this PR. So for now there's a new table `service_network_interface` which is basically the same except instead of pointing to the `instance` table it points to the `service` table.